### PR TITLE
Feat/add signup component

### PR DESCRIPTION
**Pull request by klavenjones**
Feature ticket: https://trello.com/c/pqfUazbT

- Add Signup component and Tests
- Add AuthService and ApiClientWrapper 
- Add some styling changes

This was a PR split from PR #3  
 
**NOTE**: I know it was suggested to use the Testing Library, but it wasn't easy to write some tests with it using the Vuetify library. I decided to stay with Vitest due to the ease of testing the components.

### CODE REVIEW

File: src/api/ApiClient.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  private client: AxiosInstance;
+
+  protected createAxiosClient(): AxiosInstance {
+    return Axios.create({
+      baseURL: 'http://localhost:3000',
+      responseType: 'json' as const,
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      timeout: 10 * 1000,
```
Comment (JeromeGoodrich) [inline src/api/ApiClient.ts:33]: What do these magic numbers mean?

File: src/api/ApiClient.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  put<Request, Response>(path: string, object: Request): Promise<Response>;
+  get<Response>(path: string): Promise<Response>;
+}
+
+export default class ApiClient implements IApiClient {
+  private client: AxiosInstance;
+
+  protected createAxiosClient(): AxiosInstance {
+    return Axios.create({
+      baseURL: 'http://localhost:3000',
```
Comment (JeromeGoodrich) [inline src/api/ApiClient.ts:28]: How would this work in different environments?

File: src/api/ApiClient.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+  async get<Response>(path: string): Promise<Response> {
+    try {
+      const response = await this.client.get<Response>(path);
+      return response.data;
+    } catch (error) {
+      throw new Error('Something went wrong');
+    }
+  }
+}
```
Comment (JeromeGoodrich) [inline src/api/ApiClient.ts:77]: Can you think of a way to test this file?

File: src/api/auth/AuthService.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+export interface IAuthApiClient {
+  registerUser(username: string, password: string): Promise<any>;
+}
+
+export class AuthApiClient implements IAuthApiClient {
+  baseUrl: string;
+  authApiClient: IApiClient;
+
+  constructor(authApiClient: IApiClient) {
+    this.baseUrl = 'http://localhost:3000';
```
Comment (JeromeGoodrich) [inline src/api/auth/AuthService.ts:12]: Seems like your API client has already been configured with a base URL. Do you need this one?

File: src/api/auth/AuthService.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  async registerUser(username: string, password: string): Promise<any> {
+    try {
+      const response = await this.authApiClient.post(`${this.baseUrl}/auth/register`, {
+        username: username,
+        password: password,
+      });
+
+      return response;
+    } catch (error) {
+      console.log(error);
```
Comment (JeromeGoodrich) [inline src/api/auth/AuthService.ts:25]: Is all we want to do here console log?

File: src/api/ApiClient.ts (hunk 1 above)
Comment (klavenjones) [reply src/api/ApiClient.ts]: It represents the number of milliseconds that need to be set for the request to time out. This represents milliseconds; I'll add this to a variable called milliseconds.

File: src/api/ApiClient.ts (hunk 2 above)
Comment (klavenjones) [reply src/api/ApiClient.ts]: It would be a better option to add an Environment variable as the baseURL.

File: src/api/ApiClient.ts (hunk 1 above)
Comment (JeromeGoodrich) [reply src/api/ApiClient.ts]: That's a good start, as a rule of thumb if you're not using the default, adding a breadcrumb comment on the PR and sometimes even in code can be useful

File: src/mocks/mockRegisterResponse.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export const mockRegisterResponse = {
```
Comment (erichuiwong) [inline src/mocks/mockRegisterResponse.ts:1]: Is this used anywhere?

File: src/components/__tests__/SignUp.vue.spec.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
+import { describe, it, expect, beforeEach } from 'vitest';
```
Comment (erichuiwong) [inline src/components/__tests__/SignUp.vue.spec.ts:1]: Nice tests!

File: src/components/SignUp.vue [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+          type="password"
+          required
+          :rules="passwordRules"
+          error-count="3"
+        />
+        <v-btn type="submit" data-testid="submit-button" color="primary">Sign Up</v-btn>
+      </v-form>
+    </v-card-text>
+  </v-card>
+  <h1 class="mt-6">Message: {{ message }}</h1>
```
Comment (erichuiwong) [inline src/components/SignUp.vue:20]: Do sign up pages usually have this message field?

File: src/api/ApiClient.ts (hunk 3 above)
Comment (klavenjones) [reply src/api/ApiClient.ts:65]: Yes created a mock with Axios in the test file and tested whether or not a function was called. I needed to refactor the class to inject the client this file in stead of creating a new instance when I use this class.

File: src/api/auth/AuthService.ts (hunk 5 above)
Comment (klavenjones) [reply src/api/auth/AuthService.ts]: No we should throw an error

File: src/mocks/mockRegisterResponse.ts (hunk 6 above)
Comment (klavenjones) [reply src/mocks/mockRegisterResponse.ts:1]: Yes in my test file that mocks the ApiClient

File: src/components/SignUp.vue (hunk 8 above)
Comment (klavenjones) [reply src/components/SignUp.vue:20]: No, this is just the happy path response to add my functionality.

File: src/api/__tests__/ApiClient.spec.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import { vi, expect, describe, beforeEach, it } from 'vitest';
+import ApiClient from '../ApiClient';
+import axios from 'axios';
+import { mockRegisterResponse } from '@/mocks/mockRegisterResponse';
+import { flushPromises } from '@vue/test-utils';
+
```
Comment (JeromeGoodrich) [inline src/api/__tests__/ApiClient.spec.ts:5]: Don't like seeing these warnings.

File: src/api/__tests__/ApiClient.spec.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  it('should should call axios patch method', async () => {
+    vi.spyOn(axios, 'patch').mockResolvedValueOnce(mockRegisterResponse);
+    const apiClient = new ApiClient(axios);
+
+    await apiClient.patch('/test', { data: 'Some Data' });
+
+    expect(axios.patch).toBeCalledTimes(1);
+    expect(axios.patch).toHaveBeenCalledWith('/test', { data: 'Some Data' });
+    expect(axios.patch).toHaveReturnedWith(mockRegisterResponse);
+  });
```
Comment (JeromeGoodrich) [inline src/api/__tests__/ApiClient.spec.ts:51]: What about tests for if there's an error?

**Comment by JeromeGoodrich** [comment]
Please add a link to the ticket this PR relates to in the PR description, You might also consider a PR description template to ensure you're providing ample context.

File: src/api/__tests__/ApiClient.spec.ts (hunk 10 above)
Comment (klavenjones) [reply src/api/__tests__/ApiClient.spec.ts:51]: I sent a new commit with the new tests in this PR

**Pull request merged**